render.c

```diff
@@ -93,40 +117,71 @@ cairo_surface_t *render(struct grim_state *state, struct grim_box *geometry,
-			buffer->stride);
-		cairo_pattern_t *output_pattern =
-			cairo_pattern_create_for_surface(output_surface);
-
-		// All transformations are in pattern-local coordinates
-		cairo_matrix_t matrix;
-		cairo_matrix_init_identity(&matrix);
-		cairo_matrix_translate(&matrix,
-			(double)output->geometry.width / 2,
-			(double)output->geometry.height / 2);
-		cairo_matrix_scale(&matrix,
-			(double)raw_output_width / output_width,
-			(double)raw_output_height / output_height * output_flipped_y);
-		cairo_matrix_rotate(&matrix, -get_output_rotation(output->transform));
-		cairo_matrix_scale(&matrix, output_flipped_x, 1);
-		cairo_matrix_translate(&matrix,
-			-(double)output_width / 2,
-			-(double)output_height / 2);
-		cairo_matrix_translate(&matrix, -output_x, -output_y);
-		cairo_matrix_scale(&matrix, 1 / scale, 1 / scale);
-		cairo_pattern_set_matrix(output_pattern, &matrix);
-
-		cairo_pattern_set_filter(output_pattern, CAIRO_FILTER_BEST);
-
-		cairo_set_source(cairo, output_pattern);
-		cairo_pattern_destroy(output_pattern);
-
-		cairo_paint(cairo);
-
-		cairo_surface_destroy(output_surface);
+		pixman_image_t *output_image = pixman_image_create_bits(
+			pixman_fmt, buffer->width, buffer->height,
+			buffer->data, buffer->stride);
+		if (!output_image) {
+			fprintf(stderr, "Failed to create image\n");
+			return NULL;
+		}
+		pixman_image_set_filter(output_image, PIXMAN_FILTER_BEST, NULL, 0);
+
+		// The transformations here are in reverse order; their composition
+		// sends a pixel in the output_image to one in the common_image
+		struct pixman_f_transform out2com;
+		pixman_f_transform_init_identity(&out2com);
+		pixman_f_transform_translate(&out2com, NULL,
+			-(double)output->geometry.width / 2,
+			-(double)output->geometry.height / 2);
+		pixman_f_transform_scale(&out2com, NULL,
+			(double)output_width / raw_output_width,
+			(double)output_height * output_flipped_y / raw_output_height);
+		pixman_f_transform_rotate(&out2com, NULL,
+			round(cos(get_output_rotation(output->transform))),
+			round(sin(get_output_rotation(output->transform))));
+		pixman_f_transform_scale(&out2com, NULL, output_flipped_x, 1);
+		pixman_f_transform_translate(&out2com, NULL,
+			(double)output_width / 2,
+			(double)output_height / 2);
+		pixman_f_transform_translate(&out2com, NULL, output_x, output_y);
+		pixman_f_transform_scale(&out2com, NULL, scale, scale);
+
+		struct grim_box composite_dest;
+		bool grid_aligned;
+		compute_composite_region(&out2com, buffer->width,
+			buffer->height, &composite_dest, &grid_aligned);
+
+		pixman_f_transform_translate(&out2com, NULL,
+			-composite_dest.x, -composite_dest.y);
+
+		struct pixman_f_transform com2out;
+		pixman_f_transform_invert(&com2out, &out2com);
+		struct pixman_transform c2o_fixedpt;
+		pixman_transform_from_pixman_f_transform(&c2o_fixedpt, &com2out);
+		pixman_image_set_transform(output_image, &c2o_fixedpt);
+
+		bool overlapping = false;
+		struct grim_output *output2;
```

Comment:
Nit: maybe `other_output` is a better name

Reply:
Agreed.